### PR TITLE
HACK week: Add coupon to Order [2/2]

### PR DESCRIPTION
**Pull request by selanthiraiyan**
Part 2/2 of HACK week project: Add coupon to Order
- [x] ⚠️  Make sure to review and merge part 1 #9089
- [x] ⚠️  Log GH issue to track pending work before merging this PR @selanthiraiyan https://github.com/woocommerce/woocommerce-ios/issues/9111

## Description

This PR adds an option to add a coupon to order during create/update flows

## Testing instructions

- Launch the app and login if needed
- Ensure that your site has available coupons

**Create new order**
- Navigate to `Orders` tab and tap on tap on `+`
- Tap on `+ Add Product` and add a product
- Tap on `+ Add Coupon` to open coupon screen
- Enter a valid coupon code and tap `Done` 
- Verify that you can successfully create an Order with a coupon code 
- Ensure that you see the following analytics events in Xcode console
```
Tracked order_coupon_add, properties: [AnyHashable("flow"): "creation"]
```

**Update existing order**
- Navigate to `Orders` tab and open an existing order with coupon
- Tap `Edit` on the top right to open the edit order screen
- Tap on `Coupon` row to edit the coupon
- In the textfield remove the text and enter a valid coupon code
- Tap `Done` and verify that you can successfully update a coupon to an order

**Remove coupon**
- Navigate to `Orders` tab and open an existing order with coupon
- Tap `Edit` on the top right to open the edit order screen
- Tap on `Coupon` row to edit the coupon
- Tap `Remove Coupon from Order` to delete the coupon
- Verify that the coupon is removed successfully from the order
- Ensure that you see the following analytics events in Xcode console
```
Tracked order_coupon_remove, properties: [AnyHashable("flow"): "editing"]
```

**Test feature flag**
- Disable the `addCouponToOrder` feature flag by returning `false` [here](https://github.com/woocommerce/woocommerce-ios/pull/9089/files#diff-26c9afc49bb1f1a168517d43d5e36a64ff45adb046469369ef3e4d6800ccc19bR87)
- Navigate to `Orders` tab and tap on tap on `+`
- Ensure that you can't see `+ Add a coupon` row
- Go back and navigate to the `Orders` tab and open an existing order with coupon
- Tap `Edit` on the top right to open the edit order screen
- Ensure that you can't see the `Coupon` row 

### Known issue
- We need to show a loading indicator and disable further editing when a coupon is deleted
- To know when a coupon has been deleted, we use this code [now](https://github.com/woocommerce/woocommerce-ios/pull/9090/files#diff-0667e76ef83e3a4ed8c121e825fa2a1fcc649011939b3755888c8216d1fa8c0eR450-R456)
- I reached out to @hichamboushaba to validate this, and he helpfully pointed out that it is not safe to assume that an order’s `discountTotal` will match the total of coupon line’s `discount`s. Internal slack here - p1678280360582379/1678264651.362909-slack-C03L1NF1EA3
- Since this feature is behind a flag, I will work on a fix before we release this. 
- I will log a GH issue before I merge this PR.

## Screenshots
https://user-images.githubusercontent.com/524475/223976463-c59f777d-f18e-4190-8c96-ea1dd3559bfa.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 1]
```diff
@@ -764,10 +799,16 @@ private extension EditableOrderViewModel {
                                             feeLineTotal: order.fees.first?.total ?? "0",
                                             taxesTotal: order.totalTax.isNotEmpty ? order.totalTax : "0",
                                             orderTotal: order.total.isNotEmpty ? order.total : "0",
+                                            shouldShowCoupon: order.coupons.first?.code.isNotEmpty ?? false,
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: ❓ Maybe we should add a comment explaining why we need to check for non-empty coupon code instead of non-empty coupon array.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift [hunk 2]
```diff
@@ -969,6 +1010,26 @@ private extension EditableOrderViewModel {
...
+    /// - Returns: Coupon codes comma separated to display as a summary
+    func summarizeCoupons(from couponLines: [OrderCouponLine]) -> String? {
+        guard couponLines.isNotEmpty else {
+            return nil
+        }
+
+        let output = couponLines.reduce("") { (output, line) in
+            let prefix = output.isEmpty ? "" : ","
+            return output + prefix + line.code
+        }
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: 👀 This results in coupon codes being separated by commas without space. We should add space behind the comma `", "` instead. This can also be simplified like so:

```suggestion
        let output = String(couponLines.map { $0.code }.joined(by: ", "))
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+///
+struct CouponLineDetails: View {
+
+    /// View model to drive the view content
+    ///
+    @ObservedObject private var viewModel: CouponLineDetailsViewModel
+
+    /// Defines if the coupon code input text field should be focused. Defaults to `true`
+    ///
+    @State private var focusCouponCodeInput: Bool = true
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift]: 💡 We should use `@FocusedState` instead, which helps handle the focused state nicely.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                                    .onTapGesture {
+                                        focusCouponCodeInput = true
+                                    }
+                            }
+                        }
+                        .frame(minHeight: Layout.rowHeight)
+                        .padding([.leading, .trailing], Layout.rowPadding)
+                    }
+                    .background(Color(.listForeground(modal: false)))
+                    .addingTopAndBottomDividers()
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift]: 💡 We can use the reusable view `TitleAndTextFieldRow` here to reduce the effort in the implementation.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                    Spacer(minLength: Layout.sectionSpacing)
+
+                    if viewModel.isExistingCouponLine {
+                        Section {
+                            Button(Localization.remove) {
+                                viewModel.didSelectSave(nil)
+                                presentation.wrappedValue.dismiss()
+                            }
+                            .padding()
+                            .frame(maxWidth: .infinity, alignment: .center)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift]: ❓ Is the `maxWidth` used here to make the whole row tappable? I still can't tap the button without touching the text. The `alignment` here seems redundant too.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                            .foregroundColor(Color(.error))
+                            .padding(.horizontal, insets: safeAreaInsets)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift:52]: ❓ Is this padding necessary?

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+                            .addingTopAndBottomDividers()
+                        }
+                        .background(Color(.listForeground(modal: false)))
+                    }
+                }
+            }
+            .background(Color(.listBackground))
+            .ignoresSafeArea(.container, edges: [.horizontal, .bottom])
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift:60]: ❓ Why do we have to ignore safe area on this screen? It seems to me that some text contents are outside of the safe area:

image

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetailsViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+         didSelectSave: @escaping ((OrderCouponLine?) -> Void)) {
+        self.isExistingCouponLine = isExistingCouponLine
+        self.code = code
+        self.initialCode = code
+        self.didSelectSave = didSelectSave
+    }
+
+    func saveData() {
+        let couponLine = OrderFactory.newOrderCouponLine(code: code)
+        didSelectSave(couponLine)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetailsViewModel.swift:41]: ❓ Should we validate if the coupon is correct? i.e., by fetching the coupon with the code?

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/CouponInputTransformer.swift [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+                if index == 0 {
+                    return nil
+                }
+                return OrderFactory.newOrderCouponLine(code: line.code)
+            }.compactMap({ $0 })
+            return order.copy(coupons: updatedLines)
+        }
+
+        // If there is no existing coupon lines, we insert the input one.
+        guard order.coupons.count > 0 else {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/CouponInputTransformer.swift]: Nit: let's check for not empty

```suggestion
        guard order.coupons.isNotEmpty else {
```

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/CouponInputTransformer.swift [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import Foundation
+import Yosemite
+
+/// Helper to update an `order` given an `OrderCouponLine` input type.
+///
+struct CouponInputTransformer {
+
+    /// Adds, deletes, or updates a coupon line input into an existing order.
+    ///
+    static func update(input: OrderCouponLine?, on order: Order) -> Order {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/CouponInputTransformer.swift]: Since we are updating only the first line, maybe it's worth it to add a comment regarding how it works here?

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizerTests.swift [hunk 11]
```diff
@@ -19,9 +20,17 @@ class RemoteOrderSynchronizerTests: XCTestCase {
 
     override func setUp() {
         super.setUp()
+
+        WordPressAuthenticator.initializeAuthenticator()
```
Comment (itsmeichigo) [inline WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizerTests.swift:24]: ❓ Why is this necessary?

**Review by itsmeichigo — APPROVED** [review]
Thank you @selanthiraiyan for working on this! I assume this is still a WIP so I'm approving the PR as the feature described in this PR works as expected.

A few suggestions for future improvements:
- It would be nice if we could allow users to add more than one coupon code.
- The coupon adding screen is not very user-friendly, could we show them a list of coupons that they can select from instead of entering coupon codes manually?
- When an invalid/expired/ineligible coupon code is applied, the order form displays a generic error notice saying the order cannot be updated. It would be nice if we could validate the coupon code and/or show more specific errors.
- As mentioned in this PR, the check for coupon removal is not reliable. Could we compare the coupon codes between the local order and the remote one instead?

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 1 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: Changed this to checking non-empty coupon array in d6a4eaa511b573bf82d63846805796a6a2148b7b

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift (hunk 2 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift]: Good idea. Done in 7839f79977d0f473caace4b8141ca9769f2b8eb3

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift (hunk 3 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift]: Done in ed073e86e83d7ae2bf41e79a20799dc7d5888e54

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift (hunk 5 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift]: Done in 084649402a5cf1d97a91562c41c9525d15a01062

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift (hunk 4 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift]: Done in cb2aefd9b2899378fac111d517d85919bd657527

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift (hunk 7 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift:60]: I have rearranged the views and padding to match the shipping line screen. The view looks like the following now. 

image

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetailsViewModel.swift (hunk 8 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetailsViewModel.swift:41]: Good point. 
Noted in https://github.com/woocommerce/woocommerce-ios/issues/9111

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/CouponInputTransformer.swift (hunk 9 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/CouponInputTransformer.swift]: Done in 86db93f3abf2b08895ee51b943afedb1647f80c6

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/CouponInputTransformer.swift (hunk 10 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/Synchronizer/CouponInputTransformer.swift]: Done in ef6578f0732f502080ad3643c2f63906b9393393

File: WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizerTests.swift (hunk 11 above)
Comment (selanthiraiyan) [reply WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/Synchronizer/RemoteOrderSynchronizerTests.swift:24]: When we run `RemoteOrderSynchronizerTests` tests separately (during development) without initializing `WordPressAuthenticator` it crashes.

**Comment by selanthiraiyan** [comment]
Thanks for the review and ideas, Huong! 🙏 

I have logged https://github.com/woocommerce/woocommerce-ios/issues/9111 to track the pending items from this hack week project. I will address these during the next hack week.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift]: @selanthiraiyan Just a note: we can also use a `Bool` with `@FocusedState` for simplicity if the screen has only one field. To differentiate different fields, we can use the enum solution you are using.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift (hunk 7 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift:60]: Ah got it - I think we want to ignore the safe area to make the list items expand to both the horizontal edges.

File: WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift (hunk 3 above)
Comment (selanthiraiyan) [reply WooCommerce/Classes/ViewRelated/Orders/Order Creation/PaymentSection/CouponLineDetails.swift]: Oh, I understand. Thanks for the tip. 👍
